src/psyclone/profiler.py

```diff
@@ -250,3 +253,133 @@ def gen_c_code(self, indent=0):
         '''
         raise NotImplementedError("Generation of C code is not supported "
                                   "for profiling")
+
+    def update(self):
+        '''
+        Update the underlying fparser2 parse tree to implement the profiling
+        region represented by this Node. This involves adding the necessary
+        module use statement as well as the calls to the profiling API.
+
+        :raises NotImplementedError: if the routine which is to have \
+                             profiling added to it does not already have a \
+                             Specification Part (i.e. some declarations).
+        :raises InternalError: if we fail to find the node in the parse tree \
+                             corresponding to the end of the profiling region.
+
+        '''
+        from fparser.common.sourceinfo import FortranFormat
+        from fparser.common.readfortran import FortranStringReader
+        from fparser.two.utils import walk_ast
+        from fparser.two import Fortran2003
+        from psyclone.psyGen import object_index, Schedule, InternalError
+        from psyclone.transformations import TransformationError
+
+        # Ensure child nodes are up-to-date
+        super(ProfileNode, self).update()
+
+        # Get the parse tree of the routine containing this region
+        ptree = self.root.invoke._ast
+        routines = walk_ast([ptree], [Fortran2003.Main_Program,
+                                      Fortran2003.Subroutine_Stmt,
+                                      Fortran2003.Function_Stmt])
+        for routine in routines:
+            names = walk_ast([routine], [Fortran2003.Name])
+            routine_name = str(names[0]).lower()
+            break
+
+        spec_parts = walk_ast([ptree], [Fortran2003.Specification_Part])
+        if not spec_parts:
+            # This limitation will be removed when we use the Fortran
+            # backend of the PSyIR (#435)
+            raise NotImplementedError(
+                "Addition of profiling regions to routines without any "
+                "existing declarations is not supported and '{0}' has no "
+                "Specification-Part".format(routine_name))
+        spec_part = spec_parts[0]
+
+        # Get the existing use statements
+        use_stmts = walk_ast(spec_part.content, [Fortran2003.Use_Stmt])
+        mod_names = []
+        for stmt in use_stmts:
+            mod_names.append(str(stmt.items[2]).lower())
+
+        # If we don't already have a use for the profiling module then
+        # add one.
+        if self.fortran_module not in mod_names:
+            reader = FortranStringReader(
+                "use profile_mod, only: ProfileData, ProfileStart, ProfileEnd")
```

Comment:
Ideally yes but this will all change once we have #435 - we can then use the SymbolTable. Therefore, I've adopted for a more short-term solution of checking for clashes and raising `NotImplementedError` if any are found.

Reply:
Thanks. If I were being a stickler I would say that the variable should use the namespace manager to guarantee a unique name (and this could even be done for the module names) but I think that avoiding clashes is good enough.